// smooch-core/lib/smooch.js
'use strict';

exports.__esModule = true;
exports.Smooch = exports.SERVICE_URL = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _auth = require('./utils/auth');

var _appUsers = require('./api/appUsers');

var _conversations = require('./api/conversations');

var _stripe = require('./api/stripe');

var _package = require('../package.json');

var _package2 = _interopRequireDefault(_package);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SERVICE_URL = exports.SERVICE_URL = 'https://api.smooch.io/v1';

var Smooch = exports.Smooch = function Smooch() {
  var auth = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
  var serviceUrl = arguments.length <= 1 || arguments[1] === undefined ? SERVICE_URL : arguments[1];
  (0, _classCallCheck3.default)(this, Smooch);

  this.VERSION = _package2.default.version;
  this.serviceUrl = serviceUrl;

  if (auth.keyId || auth.secret) {
    throw new Error('Key Id or Secret should not be used on the browser side. You must generate a JWT beforehand.');
  }

  this.authHeaders = (0, _auth.getAuthenticationHeaders)(auth);

  this.appUsers = new _appUsers.AppUsersApi(this.serviceUrl, this.authHeaders);
  this.conversations = new _conversations.ConversationsApi(this.serviceUrl, this.authHeaders);
  this.stripe = new _stripe.StripeApi(this.serviceUrl, this.authHeaders);

  this.utils = {};
};
